Reference for stage 3 atomics-microwait (#38316)
mdn/content: files/en-us/web/javascript/reference/global_objects/atomics/pause/index.md

---
title: Atomics.pause()
slug: Web/JavaScript/Reference/Global_Objects/Atomics/pause
page-type: javascript-static-method
browser-compat: javascript.builtins.Atomics.pause
---

{{JSRef}}

> The **`Atomics.pause()`** static method provides a micro-wait primitive, by hinting the CPU that it is spinning while waiting on a value. This releases shared CPU resources to other cores, without yielding the current thread. It has no observable behavior other than timing. The exact behavior is dependent on the CPU architecture and the operating system; for example, in Intel x86, it may be a `pause` instruction as per [Intel's optimization manual](https://www.intel.com/content/www/us/en/content-details/671488/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html). It could be a no-op in certain platforms.

Reviewer: ```suggestion
The **`Atomics.pause()`** static method provides a micro-wait primitive by hinting to the CPU that it is spinning while waiting on a value. This releases shared CPU resources to other cores without yielding the current thread. 

`pause()` has no observable behavior other than timing. The exact behavior is dependent on the CPU architecture and the operating system. For example, in Intel x86, it may be a `pause` instruction as per [Intel's optimization manual](https://www.intel.com/content/www/us/en/content-details/671488/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html). It could be a no-op in certain platforms.
```

## Syntax

```js-nolint
Atomics.pause()
Atomics.pause(durationHint)
```

### Parameters

- `durationHint` {{optional_inline}}

>   - : An integer that is positively correlated with the wait duration. Larger values result in longer waits (and more `pause` instructions sent), but the exact number has no physical meaning. There may be internal upper bound on the maximum amount of time paused on the order of tens to hundreds of nanoseconds. The hint may be ignored.

Reviewer: ```suggestion
  - : An integer that is positively correlated with the wait duration. Larger values result in longer waits (and more `pause` instructions sent), but the exact number has no physical meaning. There may be an internal upper bound on the maximum amount of time paused, to the order of tens to hundreds of nanoseconds. The hint may be ignored.
```

Reviewer: Follow-up question — what is the behavior if no `durationHint` is specified? How long is the pause?

Reviewer: Further follow-up: From reading your example explanations below, it sounds like the exact amount of time paused doesn't matter so much. You should perhaps add some explanation of this and the purpose/use case of `pause()` up near the top of the article, otherwise, this all sounds weirdly vague.

Author: > what is the behavior if no `durationHint` is specified? How long is the pause?

The spec doesn't say. I would assume it would be the same as `0`, but @syg would know better.

### Return value

None ({{jsxref("undefined")}}).

### Exceptions

- {{jsxref("TypeError")}}

>   - : Thrown if `duration` is not an integer or `undefined`.

Reviewer: ```suggestion
  - : Thrown if `durationHint` is not an integer or `undefined`.
```

## Examples

### Using Atomics.pause()

Consider a thread that tries to acquire a lock in a futex. Without micro-waits, one would just call {{jsxref("Atomics.wait()")}} or {{jsxref("Atomics.waitAsync()")}}. However, this can be inefficient because it causes the thread to be scheduled out and back in, which is expensive. Instead, we can implement a fast path by making the CPU spin for a short while to see if it can acquire the lock without waiting, and only yield back to the kernel if it can't. This is known as [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting), or [spinlocking](https://en.wikipedia.org/wiki/Spinlock).

```js
// Imagine another thread also has access to this shared memory
const sab = new SharedArrayBuffer(1024);
const i32 = new Int32Array(sab);

// Fast path: spin the CPU for a short while
let spin = 0;
do {
  if (Atomics.compareExchange(i32, 0, 0, 1) === 0) {
    break;
  }
  Atomics.pause();
  spin++;
} while (spin < 10);

// Slow path: wait for the lock
// This can only be called in a worker thread,
// because the main thread cannot be blocked
Atomics.wait(i32, 0, 1);
```

### Backoff strategies

> The `durationHint` parameter can be used to implement backoff strategies. For example, a thread can start with a small hint and increase it exponentially on each iteration. This is preferable to calling `pause` many times, because in un-JITed code, function calls themselves have a high overhead.

Reviewer: ```suggestion
The `durationHint` parameter can be used to implement backoff strategies. For example, a thread can start with a small hint and increase it exponentially on each iteration. This is preferable to calling `pause()` many times because in un-JITed code, function calls themselves have a high overhead.
```

---
title: Atomics.pause()
slug: Web/JavaScript/Reference/Global_Objects/Atomics/pause
page-type: javascript-static-method
browser-compat: javascript.builtins.Atomics.pause
---

{{JSRef}}

> The **`Atomics.pause()`** static method provides a micro-wait primitive by hinting to the CPU that it is spinning while waiting on a value. This releases shared CPU resources to other cores without yielding the current thread.

Reviewer 2: Same point as https://github.com/mdn/content/pull/38316/files#r1978656342

Author: ```suggestion
The **`Atomics.pause()`** static method Provides a micro-wait primitive that hints to the CPU that the caller is spinning while waiting on access to a shared resource. This allows the system to reduce the resources allocated to the core (such as power) or thread, without yielding the current thread.
```

`pause()` has no observable behavior other than timing. The exact behavior is dependent on the CPU architecture and the operating system. For example, in Intel x86, it may be a `pause` instruction as per [Intel's optimization manual](https://www.intel.com/content/www/us/en/content-details/671488/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html). It could be a no-op in certain platforms.

## Syntax

```js-nolint
Atomics.pause()
Atomics.pause(durationHint)
```

### Parameters

- `durationHint` {{optional_inline}}

>   - : An integer that is positively correlated with the wait duration. Larger values result in longer waits (and more `pause` instructions sent), but the exact number has no physical meaning. There may be an internal upper bound on the maximum amount of time paused, to the order of tens to hundreds of nanoseconds. The hint may be ignored.

Author: ```suggestion
  - : An integer that is positively correlated with the wait duration. Larger values result in longer waits (and more `pause` instructions sent), but the exact number has no physical meaning. There may be an internal upper bound on the maximum amount of time paused, to the order of tens to hundreds of nanoseconds. The hint may be ignored. This allows implementation of a backoff strategy by controlling the relative pause duration within each spin-loop iteration without subjecting to the overhead of JavaScript function calls; see [backoff strategies](#backoff_strategies).
```

### Return value

None ({{jsxref("undefined")}}).

### Exceptions

- {{jsxref("TypeError")}}
  - : Thrown if `durationHint` is not an integer or `undefined`.

## Examples

### Using Atomics.pause()

> Consider a thread that tries to acquire a lock in a futex. Without micro-waits, one would just call {{jsxref("Atomics.wait()")}} or {{jsxref("Atomics.waitAsync()")}}. However, this can be inefficient because it causes the thread to be scheduled out and back in, which is expensive. Instead, we can implement a fast path by making the CPU spin for a short while to see if it can acquire the lock without waiting, and only yield back to the kernel if it can't. This is known as [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting), or [spinlocking](https://en.wikipedia.org/wiki/Spinlock).

Reviewer 2: - I find this a little incomplete in that it doesn't quite convey why the fast path provides a more efficient approach or why switching in/out of the kernel. The explainer does this a bit better by capturing that 
  - the polling (spinlocking) for a short while rather than yielding is more efficient when contention is low, because you're likely to get the mutex
  - The pause creates an efficient pause, because it tells the CPU that you're not actually doing very much, allowing better use of resources. 

  >  This algorithm is both fast when contention is low and efficient when contention is high. When contention is low (i.e. the likelihood that the lock is released after a very short amount of time is high), spinning for a little bit improves the performance because the code does not re-enter the kernel to yield or to sleep. When contention is high (i.e. the likelihood that the lock is released after a very short amount of time is low), going back to the kernel to put the executing thread to sleep improves efficiency.

- It implies that busy waiting/spin locking is the whole algorithm - while it is just the bit where you poll to check if you have access (ie. it is the fast path).
- It doesn't define spinning
-  I also don't know what a futex is, though readers of this might. I'm not sure if they need to. 
- Fast path and slow path are not useful terms IMO. I would just refer to them as spinlocking and waiting.

Something like:

```suggestion
Calling {{jsxref("Atomics.wait()")}} or {{jsxref("Atomics.waitAsync()")}} in order to wait for access to shared memory causes the thread to be scheduled out of the core and then back in again after the wait.
This is efficient during times of high congestion, where access to the shared memory could take some time.
When congestion is low then it is often more efficient to poll on the lock without yielding the thread: this approach is known as [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting) or [spinlocking](https://en.wikipedia.org/wiki/Spinlock).
The `pause()` method allows you to spinlock more efficiently while waiting, by providing a hint to the CPU about what the thread is doing, and hence its likely need for resources.

In order to cater for both conditions, a common approach is to first spinlock in the hope that congestion is low, and then wait if the lock is not gained after a short time.

The example below shows how this approach can be used with `pause()` and `wait()`
```

Feel free to play with the text - I think my points are sound though

---
title: Atomics.pause()
slug: Web/JavaScript/Reference/Global_Objects/Atomics/pause
page-type: javascript-static-method
browser-compat: javascript.builtins.Atomics.pause
---

{{JSRef}}

The **`Atomics.pause()`** static method provides a micro-wait primitive that hints to the CPU that the caller is spinning while waiting on access to a shared resource. This allows the system to reduce the resources allocated to the core (such as power) or thread, without yielding the current thread.

`pause()` has no observable behavior other than timing. The exact behavior is dependent on the CPU architecture and the operating system. For example, in Intel x86, it may be a `pause` instruction as per [Intel's optimization manual](https://www.intel.com/content/www/us/en/content-details/671488/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html). It could be a no-op in certain platforms.

## Syntax

```js-nolint
Atomics.pause()
Atomics.pause(durationHint)
```

### Parameters

- `durationHint` {{optional_inline}}

>   - : An integer that is positively correlated with the wait duration. Larger values result in longer waits (and more `pause` instructions sent), but the exact number has no physical meaning. There may be an internal upper bound on the maximum amount of time paused, to the order of tens to hundreds of nanoseconds. The hint may be ignored. This allows implementation of a backoff strategy by controlling the relative pause duration within each spin-loop iteration without subjecting to the overhead of JavaScript function calls; see [backoff strategies](#backoff_strategies).

Reviewer 3: I'd suggest phrasing this as something like:

"An integer that an implementation may use to determine how long to wait. For a value `n + 1`, an implementation will wait at least as long as it does for a given value `n`. This can be used to implement a [backoff strategy] by increasing the `durationHint` passed in, although there is no guarantee that an implementation will make use of this hint."

### Return value

None ({{jsxref("undefined")}}).

### Exceptions

- {{jsxref("TypeError")}}
  - : Thrown if `durationHint` is not an integer or `undefined`.

## Examples

### Using Atomics.pause()

Calling {{jsxref("Atomics.wait()")}} or {{jsxref("Atomics.waitAsync()")}} in order to wait for access to shared memory causes the thread to be scheduled out of the core and then back in again after the wait. This is efficient during times of high contention, where access to the shared memory could take some time. When contention is low, then it is often more efficient to poll on the lock without yielding the thread: this approach is known as [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting) or [spinlocking](https://en.wikipedia.org/wiki/Spinlock). The `pause()` method allows you to spinlock more efficiently while waiting, by providing a hint to the CPU about what the thread is doing, and hence its low need for resources.

> 

Reviewer 3: It may be worth adding a note that although possible to use this from the main thread, it's not recommended, unless you are already forced to use a spinlock for some other reason.

To cater for both conditions, a common approach is to first spinlock in the hope that contention is low, and then wait if the lock is not gained after a short time. If we acquired the lock via spinlocking already, then the `wait()` call will be a no-op.

The example below shows how this approach can be used with `Atomics.pause()` and `Atomics.wait()`.

```js
// Imagine another thread also has access to this shared memory
const sab = new SharedArrayBuffer(1024);
const i32 = new Int32Array(sab);

// Fast path: spin the CPU for a short while
let spin = 0;
do {
  if (Atomics.compareExchange(i32, 0, 0, 1) === 0) {
    break;
  }
  Atomics.pause();
  spin++;
} while (spin < 10);

// Slow path: wait for the lock
// This can only be called in a worker thread,
// because the main thread cannot be blocked
Atomics.wait(i32, 0, 1);
```

### Backoff strategies

> The `durationHint` parameter can be used to implement backoff strategies. For example, a thread can start with a small hint and increase it exponentially on each iteration. This is preferable to calling `pause()` many times because in un-JITed code, function calls themselves have a high overhead.

Reviewer 3: We should probably add a note that there's no guarantee an implementation will make use of `durationHint`.
